Assets/Mirage/ObjectPolling/ObjectPoolManager.cs

```diff
@@ -0,0 +1,204 @@
+        /// <returns></returns>
+        internal NetworkIdentity NetworkSpawnPool(Vector3 position, Quaternion rotation, Guid AssetId, int quantity = 3)
+        {
+            int prefabId = _objectsAssetIds[AssetId].gameObject.GetInstanceID();
+
+            if (_recycledObjects.ContainsKey(prefabId))
+                return (_recycledObjects[prefabId] is ObjectPooling<NetworkIdentity>
+                    ? (ObjectPooling<NetworkIdentity>)_recycledObjects[prefabId]
+                    : default).NetworkSpawn(position, rotation);
+
+            var createParent = new GameObject(_objectsAssetIds[AssetId].name) { transform = { parent = _parent.transform } };
+
+            _recycledObjects[prefabId] = new ObjectPooling<NetworkIdentity>(_objectsAssetIds[AssetId], quantity, createParent);
+
+            return (_recycledObjects[prefabId] is ObjectPooling<NetworkIdentity>
+                ? (ObjectPooling<NetworkIdentity>)_recycledObjects[prefabId]
+                : default).NetworkSpawn(position, rotation);
+        }
+
+        /// <summary>
+        ///     Network object pooler unspawning of object and resetting parent back
+        ///     to the object pooler.
+        /// </summary>
+        /// <param name="objectSpawned">The object we want to set in active and despawn.</param>
+        /// <param name="isServer">Whether or not this is server telling us to unspawn object.</param>
+        /// <param name="isEnemy">Whether or not this is an enemy being despawn-ed.</param>
+        internal void NetworkUnSpawnObject(NetworkIdentity objectSpawned, bool isServer, bool isEnemy)
+        {
+            UnSpawnObject(objectSpawned);
+
+            // If this is network server then tell rest of clients to unspawn object.
+            if (_serverObjectManager == null || !isServer) return;
+
+            // Unspawn object.
+            _serverObjectManager.Destroy(objectSpawned.gameObject, false);
+        }
+
+        #endregion
+
+        #region Local Spawning.
+
+        /// <summary>
+        ///     Spawn an object using our pooled objects for the specific object
+        ///     or create a new pool if no pool has been created not networked objects.
+        /// </summary>
+        /// <param name=""></param>
+        /// <param name="gObject">The object we want to spawn.</param>
+        /// <param name="position">The position we want to spawn the object at.</param>
+        /// <param name="rotation"></param>
+        /// <param name="worldStayPosition">Unity specific <see cref="Instantiate"/></param>
+        /// <param name="quantity">How many pooled objects we want to create of this object. If we don't have a pool started.</param>
+        /// <returns></returns>
+        public T LocalSpawnPool<T>(T gObject, Vector3 position, Quaternion rotation, bool worldStayPosition = false, int quantity = 3) where T : Object
```

Comment:
do we need these function as well as the functions above?

Reply:
Well I added them so end users can use a full blown object pooling system. People do need to use local object pooling for spawning particles and such and if we are going to provide builtin system might as well provide one for local object pooling too.